develop/dev-guide-ticdc-checksum-verification.md

```diff
@@ -0,0 +1,530 @@
+        "Accept",
+        "application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, "+
+            "application/json",
+    )
+
+    httpClient := &http.Client{}
+    resp, err := httpClient.Do(req)
+    if err != nil {
+        return nil, err
+    }
+    defer resp.Body.Close()
+
+    body, err := io.ReadAll(resp.Body)
+    if err != nil {
+        log.Error("Cannot parse the lookup schema response", zap.Error(err))
+        return nil, errors.WrapError(errors.ErrAvroSchemaAPIError, err)
+    }
+
+    if resp.StatusCode == 404 {
+        log.Warn("Specified schema not found in Registry", zap.String("requestURI", requestURI), zap.Int("schemaID", schemaID))
+        return nil, errors.ErrAvroSchemaAPIError.GenWithStackByArgs("Schema not found in Registry")
+    }
+
+    if resp.StatusCode != 200 {
+        log.Error("Failed to query schema from the Registry, HTTP error",
+            zap.Int("status", resp.StatusCode), zap.String("uri", requestURI), zap.ByteString("responseBody", body))
+        return nil, errors.ErrAvroSchemaAPIError.GenWithStack("Failed to query schema from the Registry, HTTP error")
+    }
+
+    var jsonResp lookupResponse
+    err = json.Unmarshal(body, &jsonResp)
+    if err != nil {
+        log.Error("Failed to parse result from Registry", zap.Error(err))
+        return nil, errors.WrapError(errors.ErrAvroSchemaAPIError, err)
+    }
+
+    codec, err := goavro.NewCodec(jsonResp.Schema)
+    if err != nil {
+        return nil, errors.WrapError(errors.ErrAvroSchemaAPIError, err)
+    }
+    return codec, nil
+}
+
+type lookupResponse struct {
+    Name     string `json:"name"`
+    SchemaID int    `json:"id"`
+    Schema   string `json:"schema"`
+}
+
+```
+
+## 计算并校验 Checksum
+
+上一步拿到的 `valueMap` 和 `valueSchema`，即含有用于 checksum 计算和校验的所有元素。
+
+在消费端计算和校验 Checksum 的过程可以总结为以下几个步骤：
+
+1. schema 中含有每个列的类型信息，列顺序和 Checksum 计算顺序相同。
+2. 遍历每一列，根据列的数据值和对应的 MySQL Type，生成字节切片，不断更新 Checksum。
+3. 将上一步计算得到的 Checksum 和从收到的消息里提取出来的 Checksum 做比较。如果不一致，则说明 Checksum 校验失败，数据可能发生损坏。
+
+示例代码如下：
+
+```go
+func CalculateAndVerifyChecksum(valueMap, valueSchema map[string]interface{}) error {
+    // fields 存放有数据变更事件的每一个列的类型信息，按照每一列的 ID 排序，该顺序和 Checksum 计算顺序相同
+    fields, ok := valueSchema["fields"].([]interface{})
+    if !ok {
+        return errors.New("schema fields should be a map")
+    }
+
+    // 1. 从 valueMap 里面查找期望的 checksum 值，它被编码成 string 类型
+    // 如果找不到，说明 TiCDC 发送该条数据时，还没有开启 checksum 功能，直接返回即可
+    o, ok := valueMap["_tidb_row_level_checksum"]
+    if !ok {
+        return nil
+    }
+    expected := o.(string)
+    if expected == "" {
+        return nil
+    }
+
+    // expectedChecksum 即是从 TiCDC 传递而来的期望的 checksum 值
+    expectedChecksum, err := strconv.ParseUint(expected, 10, 64)
+    if err != nil {
+        return errors.Trace(err)
+    }
+
+    // 2. 遍历每一个 field，计算 checksum 值
+    var actualChecksum uint32
+    // buf 用来存储每次更新 checksum 时使用的字节切片
+    buf := make([]byte, 0)
+    for _, item := range fields {
```

Comment:
The ordering of fields is fixed or not?

Reply:
yes, it's fixed, the same as the checksum calculation.